Jenkins fails when smoketests in test pods fail

diverdane opened the pull request:
This change fixes a problem whereby the Jenkins CI job was not being
marked as a failure when the smoketests in the test pods fail.

This appears to be a bug in the Google marketplace tools. The script
that's being used to deploy and test Conjur is
`marketplace-k8s-app-tools/scripts/verify`. The problem with this script
is that it checks whether the `deployer` job succeeds, but it doesn't
check for success/failure of the two test pods that are created.

This change also rolls back the version of Conjur OSS that is used in CI
from 1.6 to 1.5, since Conjur OSS v1.6 does not yet have the (revert) fix for
https://github.com/cyberark/conjur/pull/1499, causing one of the Helm
test pods to fail.

Fixes Issue #49

sgnn7 commented on ci/jenkins_build:
```diff
@@ -1,14 +1,31 @@
+    echo "$1"
+    echo "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~"
+    echo "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~"
+}
```
This is probably unnecessary vs the standard announce (even the announce is probably overkill)

sgnn7 commented on ci/jenkins_build:
```diff
@@ -50,9 +67,164 @@ docker run --rm \
+    fi
+
+    elapsed_time=$(( $(date +%s) - $start_time ))
+    printf "Elapsed ${elapsed_time}s\r"
```
There's way too much mixing of printing methods: `echo`, `printf`, `announce`, and `final_result` you should try to cu this list at least in half.

sgnn7 commented on ci/jenkins_build:
```diff
@@ -50,9 +67,164 @@ docker run --rm \
+      || clean_and_exit "Failed to stop application"
+
+  announce "Wait for resources to be deleted from the namespace"
+  deletion_timeout=180
```
If this is the only value we use, you can either make it a top-level constant or embed it into `wait_for_deletion` (I'm not suer if anything else uses this method).

sgnn7 commented on ci/jenkins_build:
```diff
@@ -50,9 +67,164 @@ docker run --rm \
+  final_result "                 SUCCESS"
+  display_manifests=false
+
   trap - EXIT
```
Do we have other traps? If not, this can be removed. If we also don't run programs after this in the same shell, the `trap popd` should go away too I think

diverdane replied:
I removed all of the pushd/popd and trap logic except the `trap delete_namespace EXIT`, which I think is safe programming practice?

sgnn7 commented on ci/jenkins_build:
```diff
@@ -50,9 +67,164 @@ docker run --rm \
+
+  # Wait for RBAC resources to be deleted from the namespace
+  wait_for_deletion "serviceaccounts,roles,rolebindings" "$deletion_timeout" \
+  || clean_and_exit "Some service accounts or roles were not deleted"
```
Can all of these deletions be done by deleting the namespace?

diverdane replied:
Yes. I'll remove the deletions of standard Kubernetes resources and RBAC from the namespace.
However, I would like to keep the deletion (and waiting for deletion success) for the Conjur application itself, since I think we document that step in various places, so might be good to keep a test of this.

sgnn7 commented on ci/jenkins_build:
```diff
@@ -50,9 +67,164 @@ docker run --rm \
+
+    if [[ "$res_count" -eq 0 ]]; then
+      break
+    else
```
This needs flattening:
```bash
    res_count=$(echo $resources | wc -w)
    if [[ "$res_count" -eq 0 ]]; then
      break
    fi

    # Ignore service account default
    if [[ "$resources" = "ServiceAccount/default" ]]; then
        break
    fi

    echo "INFO Remaining: $res_count"
    echo "INFO $resources"
...
```

diverdane replied:
Fixed.

diverdane commented:
@sgnn7  Thank you for your review!
I've responded to your comments and made the appropriate changes, so should be ready for re-review.
A couple of notes:
(a) I've added a `usage` function and `--help` option to the build.sh script, since in the middle of changing around variable names, I had forgotten about the `NAME` and `NAMESPACE` env variables that this script uses... so probably useful to anyone using this repository in the future to have this usage/help.
(b) I changed bash variable names so that environment variables use ALL CAPS and local variables use lower case.
